### ISSUE TITLE
Respect the order of tags in include directive

### ISSUE THREAD

**Issue by mojavelinux** [asciidoctor/asciidoctor#2368]
When inserting tagged content using the include directive, respect the order in which the tags are specified. This allows content to be arranged in an order that differs from the order the content appears in the document.

To achieve this behavior today, it's necessary to use multiple include directives.

**Comment by mojavelinux** [comment]
The same could be said for the list of line ranges.

**Comment by mojavelinux** [comment]
This turns out to be fairly tricky if we consider the fact that a tag can occur multiple times in a document and that there are wildcard filters.

We might have to narrow the scope of it. Something like: if there are no wildcards, and each tag matches exactly once in the document, then respect the order of the tags in the include directive.

**Comment by rockyallen** [comment]
If I understand the proposal properly, this is a really bad idea.
1. Logically identical documents will render differently.
Consider
````
include::doc.adoc[tags="a,b"]
````
with
````
// tag::b[]
1a
1b
// end::b[]
// tag::a[]
2
// end::a[]
````
= 2, 1a, 1b
but the (to me) logically identical
````
// tag::b[]
1a
// end::b[]
// tag::b[]
1b
// end::b[]
// tag::a[]
2
// end::a[]
````
= 1a, 1b, 2
2. It will cause confusion. What if an editor decides 1b is no longer required? he deletes that block, then doesn't understand why the document suddenly gets mangled (or worse, doesn't notice). Even if he understands, it requires him to count tags manually to know what will happen-an error prone waste of time.
3. It breaks existing documents, either annoying people or requiring more compatibility flags to switch it off.
If you really want this, could you give it a special syntax? perhaps include::doc.adoc[tags="a>b"] for "tag a then tag b", but even so I suspect the conceptual load outweighs any time saved in typing.
Or have I misunderstood?

**Comment by higgspa** [comment]
I guess It all depends on how one is actually leveraging the include/tag directive. For my use case, I like to leverage the tags for reusable content and actually assumed the tags would be included in the order I listed as that made logical sense to me. However, as I added more and more content I realized that was not the case. I then dug deeper into the tag documentation and see many of the benefits of the current design - of which I am taking advantage of.  

With that said, I would get tremendous value with an option to include tags in the order I specified. I agree that it would need to be backward compatible and not break existing documents. Perhaps a new flag that specifically states to honor the ordering.

**Comment by rockyallen** [comment]
I agree with @higgspa that both styles should be available.
But you should also be able to switch styles inside the same document, which you can't do with a flag. Using a different syntax would allow this, with the bonus of letting you throw an error on illegal combinations.

eg:

include::file[tags=a,b] // document order as now, therefore backward compatible
include::file[tags=a>b] // order respecting, with an allowable tag combination
include::file[tags=*>b] // "ERROR: incorrect tag specification. Order-respecting include cannot contain wildcard tag names."

**Comment by mojavelinux** [comment]
Keep in mind that this could be implemented using a custom include processor. What we're talking about here is whether it belongs in core...and what the rules would be.

**Comment by mojavelinux** [comment]
I agree that if we do add this, it probably has to be explicit. We certainly don't want to break existing documents.

Though, I do have to question what the use case is for listing the tags in an order different than what they are in the file. In all cases I checked, the order of tags always matches the document order. So I think most people were doing this subconsciously. (I'd like to see a document that breaks this assumption).

While I think the explicit attribute is a good idea, I'm still thinking about how this could be expressed in shorthand. We should look at where order is used in other systems. One would be graph notation. (similar to what @rockyallen suggested)

```
tags=a->b->c
```

As for the explicit attribute, when there are only two possible values, we tend to prefer an option. So it would be something like opts=ordered (it's implied that the order refers to the tags as the document order is a natural default).

**Comment by elextr** [comment]
> Though, I do have to question what the use case is for listing the tags in an order different than what they are in the file. 

The use-case I had in mind for wanting the tags in specified order was when extracting stuff from source code files for documentation purposes, to re-order it from the order enforced by programming language semantics (eg declare before first use) to the order that suited discussing it, or maybe even alphabetic if extracting API functions.  

>  In all cases I checked, the order of tags always matches the document order. 

You will probably not see any different example in an existing document since its not implemented so nobody does it. (chicken and egg)

> So it would be something like opts=ordered

Sure, custom processors can still add values without syntax change, just that `opts=` is already heavily abused and there is gonna be clashes one day.

My only problem with a short-cut syntax is its not extendable, so if you want to allow custom processors to do something else you need the parameter method as well.

My advice is to make it _possible_ first with the extendable parameter/option type syntax and see how much it is used, then if its used enough and there is a clamour for a shortcut, a syntax sugar can be added.

**Comment by mojavelinux** [comment]
> My advice is to make it possible first with the extendable parameter/option type syntax and see how much it is used, then if its used enough and there is a clamour for a shortcut, a syntax sugar can be added.

That's sound advice.

> The use-case I had in mind for wanting the tags in specified order

I totally understand why we need this feature. What I was asking is why someone would currently put tags in a different order given that it currently has no meaning. This is to address @rockyallen's concern that we would break existing documents. I don't think there would be many cases, if any.

> You will probably not see any different example in an existing document since its not implemented so nobody does it.

Exactly what I'm saying. I'm pretty sure the tags are already in document order since that's the only mode currently supported.

>  just that opts= is already heavily abused and there is gonna be clashes one day.

I'm not sure why you think opts is abused. opts is a general purpose attribute for switches. It can accommodate any arbitrary number of switches. We'd just be adding another possible entry. For example, you could have `opts="ordered,optional"` (optional being another feature of the include directive).

> My only problem with a short-cut syntax is its not extendable, so if you want to allow custom processors to do something else you need the parameter method as well.

A custom include processor takes over the entire processing of the include directive, so it can do whatever it wants. We would not be limiting what it can do in any way.

**Comment by higgspa** [comment]
My use case was related to REST API documentation. I created multiple files to hold snippets of reusable content (shared libraries if you will) to describe a REST API path segments, query parameters, JSON properties etc. When I include the tags, for a specific API's documentation, I would like them listed (documented) in the order they appear in the URL which might be a different order than another API. If this is still confusing I can work up a simple example.

**Comment by mojavelinux** [comment]
Thank you for the clarification.

I'll remind you that it's possible to use your own IncludeProcessor to implement this functionality. The reason it's difficult is because not all the functionality of the built-in include processor is compatible with ordering. Even if ordering is specified explicitly, we can still only order if the all the tag names you want to include are specified explicitly. That's why this is feeling more and more like a use case for a custom extension.

**Comment by higgspa** [comment]
Makes sense. I will look into a custom extension. Thanks.
